File: src/main/java/com/phasmidsoftware/dsaipg/adt/threesum/ThreeSumQuadratic.java
/*
 * Copyright (c) 2024. Robin Hillyard
 */

package com.phasmidsoftware.dsaipg.adt.threesum;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

/**
 * Implementation of ThreeSum which follows the approach of dividing the solution-space into
 * N sub-spaces where each sub-space corresponds to a fixed value for the middle index of the three values.
 * Each sub-space is then solved by expanding the scope of the other two indices outwards from the starting point.
 * Since each sub-space can be solved in O(N) time, the overall complexity is O(N^2).
 * <p>
 * NOTE: The array provided in the constructor MUST be ordered.
 */
public class ThreeSumQuadratic implements ThreeSum {
    /**
     * Construct a ThreeSumQuadratic on a.
     *
     * @param a a sorted array.
     */
    public ThreeSumQuadratic(int[] a) {
        this.a = a;

        length = a.length;
    }

    /**
     * Retrieves an array of unique Triples. Each Triple represents a unique combination of three integers from
     * the source array that sum to zero.
     *
     * @return an array of distinct Triples, sorted in natural order, where each Triple satisfies the condition that
     * the sum of its three integers is zero.
     */
    public Triple[] getTriples() {
        List<Triple> triples = new ArrayList<>();
        for (int i = 0; i < length; i++) triples.addAll(getTriples(i));
        Collections.sort(triples);
        return triples.stream().distinct().toArray(Triple[]::new);
    }

    /**
     * Get a list of Triples such that the middle index is the given value j.
     *
     * @param j the index of the middle value.
     * @return a Triple such that
     */
     List<Triple> getTriples(int j) {
         List<Triple> triples = new ArrayList<>();
         int left = 0;
         int right = length-1;
         while (left < j && right > j) {
             int sum = a[left] + a[j] + a[right];
             if (sum > 0) {
                 right -= 1;
             }
             else if (sum < 0) {
                 left += 1;
             }
             else {
                 triples.add(new Triple(a[left], a[j], a[right]));
                 left++;
                 right--;
                 while (left < j && left > 0 && a[left] == a[left - 1]) left++;
                 while (right > j && right < length - 1 && a[right] == a[right + 1]) right--;
             }

         }
         return triples;
    }

    private final int[] a;
    private final int length;
}

File: Python/src/adt/threesum/three_sum_quadratic.py
"""
Quadratic three-sum, ported from adt/threesum/ThreeSumQuadratic.java.
"""

from __future__ import annotations

from src.adt.threesum.three_sum import ThreeSum
from src.adt.threesum.triple import Triple


class ThreeSumQuadratic(ThreeSum):
    """
    Divides the solution space into n sub-spaces, one per middle index, and solves
    each in linear time by working outwards from it. Overall O(n^2).

    NOTE the list given to the constructor MUST be sorted and distinct. Nothing
    checks this, and an unsorted list produces a wrong answer rather than an
    error.
    """

    def __init__(self, a: list[int]) -> None:
        """
        :param a: a sorted list of distinct values.
        """
        self._a = a
        self._length = len(a)

    def get_triples(self) -> list[Triple]:
        """
        :return: the distinct Triples summing to zero, in ascending order.
        """
        triples: list[Triple] = []
        for i in range(self._length):
            triples.extend(self.get_triples_with_middle(i))
        return sorted(set(triples))

    def get_triples_with_middle(self, j: int) -> list[Triple]:
        """
        Get the Triples whose middle element is at index j.

        Two indices walk outwards from j, one down and one up. At each step the
        value needed at the top is determined by the one at the bottom, so a
        single comparison says which index to move -- which is what makes this
        sub-problem linear rather than quadratic.

        NOTE named get_triples_with_middle rather than get_triples, because Python
        cannot overload: the Java distinguishes getTriples(int) from getTriples()
        by arity.

        :param j: the index of the middle value.
        :return: the Triples with a[j] in the middle.
        """
        # TO BE IMPLEMENTED : for each candidate, test if a[i] + a[j] + a[k] = 0.
        raise NotImplementedError("TO BE IMPLEMENTED")
